Remote Control - Allow Multiple Modules per Module Type (#696)
smartdevicelink/sdl_evolution: proposals/nnnn-multiple-modules.md

# Remote Control - Allow Multiple Modules per Module Type

* Proposal: [SDL-NNNN](nnnn-multiple-modules.md)
* Author: [Zhimin Yang](https://github.com/smartdevicelink/yang1070)
* Status: **Awaiting Review**
* Impacted Platforms: [Core / iOS / Android / RPC]

## Introduction

> SDL remote control baseline proposal [SDL-0071](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0071-remote-control-baseline.md) group available controllable items of the same category into a RC module type. SDL defines six module types: `RADIO`, `CLIMATE`, `AUDIO`, `LIGHT`, `HMI_SETTINGS` and `SEAT`. SDL allows only one module per module type so far. However, there are production needs to support multiple modules of the same module type. For example, some vehicles in the market are equipped with "dual-zone" climate control systems. There are also ["tri-zone" or even "quad-zone" climate control](https://www.autotrader.com/car-info/definitions-dual-zone-climate-control-211783) in the existing vehicles. Some vehicles in the market have seat control for several seats (driver's seat, front passenger's seat for example). Some vehicles have both a default entertainment system and a back seat entertainment system. There are car headrest monitors, seat back monitors, car roof mount monitors and flip down screens. Passengers can use a back seat entertainment system to play video, audio or even games. 

Author: changed `group` to `groups`

## Motivation

> To allow SDL support multiple modules of the same module type, SDL needs a module id parameter.

Author: updated

    </param>
</struct>
```

The `ModuleId` serves two purposes. First, its mandatory parameter `id` is used as an ID helper for a RC module within a module type. Second, it has additional information about the RC module. 

`location` indicates where the module is located. 

`serviceArea` indicates the service area or the coverage area of a module. For example, a seat control covers only one seat. The location and the service area of a seat control module are the same. A radio control serves all passengers in the vehicle. Its service area will cover its own physical location as well as other locations in the vehicle. 

Flag `share` indicates whether a RC module can be shared among multiple users or exclusively used by a user. In general, examples of shared module types are `RADIO`, `AUDIO` and `LIGHT`. Examples of exclusive module types are `SEAT` and `HMI`. Depending on single climate zone or multiple climate zone support, a module with `CLIMATE` module type can be shared or exclusive.

Ideally, `ModuleID` can also include a `moduleType` and a `moduleName`. However, moving the existing mandatory parameters from other data types into the new data type will cause a breaking change.

> The new parameter `moduleID` is added to `GetSystemCapabilites` response within data type `Xyz-ControlCapabilities`. Where the `Xyz` in `Xyz-ControlCapabilities` can be any of the existing 6 module types. When an app receives the remote control capabilities, it shall remember the capabilities and it might show a map to the user indicating where a module is and a module's coverage so that the user can choose a module to control. When the app send a `GetInteriorVehicleData` or a `SetInteriorVehicleData` request, it shall use the module `id` together with the module type to identify a module. Instead of sending the whole `moduleID` data, only the `id` string is needed. Valid ids are those published by the `GetSystemCapabilites` response.

Author: updated, send -> sends

```xml
<struct name="Grid" since="5.x">
    <description>Describes a location (origin coordinates and span) of a vehicle component.</description>
    <param name="col" type="Integer" mandatory="true" minvalue="-1" maxvalue="100">
    </param>
    <param name="row" type="Integer"  mandatory="true" minvalue="-1" maxvalue="100">
    </param>
    <param name="level" type="Integer" mandatory="false"  defvalue="0" minvalue="-1" maxvalue="100" >
    </param>
    <param name="colspan" type="Integer" mandatory="false" defvalue="1" minvalue="1" maxvalue="100">
    </param>
    <param name="rowspan" type="Integer" mandatory="false" defvalue="1" minvalue="1" maxvalue="100">
    </param>
    <param name="levelspan" type="Integer" mandatory="false" defvalue="1" minvalue="1" maxvalue="100">
    </param>
</struct>
```

> For example, in a regular sedan with two rows, there columns and one level. The following image and table show the positions and the grid of each seat.

Author: yes, `there` shall be `three`

| row=1 | rear-left seat : 1A : {col=0, row=1, level=0, colspan=1, rowspan=1, levelspan=1}| rear-middle seat : 1B : {col=1, row=1, level=0, colspan=1, rowspan=1, levelspan=1} | rear-right seat : 1C :{col=2, row=1, level=0, colspan=1, rowspan=1, levelspan=1} |

Please note grids can be overlapping. A grid that covers the completely internal area of the above two-row three-column vehicle can be {col=0, row=0, colspan=3, rowspan=2}.

The following images and tables show the positions of each seat in popular 7-seat and 12-seat configurations with 3x3 and 4x4 grids.

![7 Passenger](/assets/proposals/nnnn-multiple-modules/3x3.jpg)

| col=-1, row=-1  | col=0 | col=1 | col=2 |
| --- | --- | --- | --- |
| row=0 | driver's seat: 0A  |   | front passenger's seat : 0C: |
| row=1 | 1A |  |  1C  |
| row=2 | 2A | 2B |  2C  |

![11 or 12 Passenger](/assets/proposals/nnnn-multiple-modules/4x4.jpg)

| col=-1, row=-1  | col=0 | col=1 | col=2 | col=3 
| --- | --- | --- | --- | --- | 
| row=0 | driver's seat: 0A  |  |  | front passenger's seat : 0D: |
| row=1 | 1A | 1B |  1C  | |
| row=2 | 2A | 2B |  2C  | |
| row=3 | 3A | 3B |  3C  | 3D |

>  The vehicle HMI needs to publish how many rows, columns and levels are available for the vehicle and the list of modules in some properly defined grids. This proposal propose to publish the row number, column number and level number in a new `SeatLocationCapability`, which also includes all the seats installed and their locations. An app may use this information to show a user a seat map. `grid` is mandatory when the parameter is used in a message sent from the vehicle to mobile apps. The seat `id` must be an id published by the capabilities. 

Author: updated to `proposes`

```xml
<struct name="SeatLocation" since="5.x">
    <description>Describes the location of a seat.</description>
    <param name="id" type="String" mandatory="true">

>         <description>A shot meaningful name coming from the grid in the format of `digits + letter` for single leveled vehicle or `digits + letter + digits` for multiple leveled vehicle. Example 0A, 1B, 2C0. left digits=row number, right digits=level number, letter from column number: A for 0, B for 1, C for 2 and so on. </description>

Author: updated all instance from `shot` to `short`

SDL core and HMI need to validate `moduleId` string against remote control capabilities. A valid moduleId means it is published by the HMI via remote control capabilities.

### Missing moduleId

SDL core and HMI shall treat a missing moduleId (older version API) as the default module id for the module type. It is up to the HMI to determine what moduleId is the default id for the module type.

- Remote control capabilities

HMI shall always provide a module id and service area in the capabilities when it supports multiple modules per module type. When there is only one module per module type, HMI shall use a default id (or the first item in the list) of this module type.

- GetInteriorVehicleData/SetInteriorVehicleData request and response

If the optional moduleId is provided in a `GetInteriorVehicleData`/`SetInteriorVehicleData` request, SDL core shall either forward the request to HMI if it is a valid id or reject the request if it is an invalid id. 

If the optional moduleId is not provided in a `SetInteriorVehicleData` request, SDL core shall forward the request as is (without moduleId) to HMI. HMI shall use the default id for this module type when processing the request. 

> If the optional moduleId is not provided in a `GetInteriorVehicleData` request, and if there are more than one modules (published by capabilities) of the same module type, SDL core shall forward the request as is (without moduleId) to HMI. If there is only one module published of the same module type, SDL core shall return cached data or forward the request to HMI if cached data is not available. HMI shall use the default moduleId for this module type when processing the request. 

Author: update `more than one modules` to `more than one module`

> In all cases, HMI/SDL must always include a moduleId in the response even it is optional and only change/return the data for the requested moduleId. SDL core shall cache the module data for this module id + module type and forward the response to the mobile.

Reviewer: ```suggestion
In all cases, HMI/SDL must always include a moduleId in the response even if it is optional and only change/return the data for the requested moduleId. SDL core shall cache the module data for this module id + module type and forward the response to the mobile.
```

Author: added `if`

- OnInteriorVehicleData notifications

HMI and SDL core shall always include a moduleId for a `ModuleData` in an `OnInteriorVehicleData` notification. SDL core shall update the cache.

- OnRCStatus notifications and module allocation

SDL core shall always include a moduleId for each and every `ModuleData` in `allocatedModules` and `freeModules` in an `OnRCStatus` notification.

#### Policy Update
This proposal does not propose any change to the policy.

Consider a simple policy extension that is similar to "moduleType", which adds a list of allowed module types to an app's policy, we might add a list of allowed module IDs as the example below.

```json
app_policies": {
  "xyz app": {
      ....
      "moduleType": ["RADIO", "CLIMATE"],
      "grid": ["id1", "id2", "id3"]
   }
}
```

> However, we think it is unnecessary in the current stage. When an app partner develops a remote control SDL app to control seats for example, the expectation is to be able to control all available seats by default. At run time, there might be permission pop ups or permission settings screen for the driver to give permissions to an app (potentially multiple instances of the same app) used by the users sitting in the rear seats to control their own seats. The policy server will not be able to configure that an app can access only certain seats statically.

Author: added `a`

    <element name="PHONE_CALL"/>
    <element name="VIDEO_STREAMING"/>
    <element name="REMOTE_CONTROL"/>
+   <element name="SEAT_LOCATION"/>
</enum>

<struct name="SystemCapability" since="4.5">
    ...
+    <param name="seatLocationCapability" type="SeatLocationCapability" mandatory="false" since="5.x">
+        <description>Contains information about the locations of each seat</description>
+    </param>
</struct>

<struct name="SeatLocationCapability" since="5.x">
    <description>Contains information about the locations of each seat</description>
    <param name="rows" type="Integer" minvalue="1" maxvalue="100" mandatory="false">
    <param name="columns" type="Integer" minvalue="1" maxvalue="100" mandatory="false">
    <param name="levels" type="Integer" minvalue="1" maxvalue="100" defValue="1" mandatory="false">   
    </param>
+    <param name="seats" type="SeatLocation" array="true" mandatory="false">
+        <description>Contains a list of SeatLocation in the vehicle</description>
+    </param>
</struct>

<struct name="SeatLocation" since="5.x">
    <description>Describes the location of a seat.</description>
    <param name="id" type="String" mandatory="true">

>         <description>A shot meaningful name coming from the grid in the format of `digits + letter` for single leveled vehicle or `digits + letter + digits` for multiple leveled vehicle. Example 0A, 1B, 2C0. left digits=row number, right digits=level number, letter comes from column number: A for 0, B for 1, C for 2 and so on. </description>

Reviewer: ```suggestion
        <description>A short meaningful name coming from the grid in the format of `digits + letter` for a single leveled vehicle or `digits + letter + digits` for a multiple leveled vehicle. Example 0A, 1B, 2C0. left digits=row number, right digits=level number, letter comes from column number: A for 0, B for 1, C for 2 and so on. </description>
```

Author: updated all instances

        <element name="TOO_MANY_PENDING_REQUESTS"/>
        <element name="APPLICATION_NOT_REGISTERED"/>
        <element name="GENERIC_ERROR"/>
        <element name="REJECTED"/>
        <element name="IGNORED"/>
        <element name="DISALLOWED"/>
        <element name="USER_DISALLOWED"/>
        <element name="UNSUPPORTED_RESOURCE"/>
    </param>
    <param name="info" type="String" maxlength="1000" mandatory="false">
    </param>
    <param name="success" type="Boolean" platform="documentation" mandatory="true">
        <description> true if successful; false, if failed </description>
    </param>
</function>
```

An alternative solution is to add a new RPC `SetUserLocation`, which is sent from mobile apps to SDL and forwarded to HMI, instead of including `userLocation` as a parameter of the `GetInteriorVehicleDataConsent` RPC.

### How to use the new parameter for permission control

A simple Module Id based permission control is listed here.

- SDL shall enforce permission control. It may or may not use the local policy table depending on the implementation. The permissions are given by the user/driver via HMI, and information is passed to sdl core.

> - When a user launches a RC app from a phone, the app calls GetSystemCapabilites to get vehicle configurations, which includes how many rows, columns and levels are available, how many / what type / what id of remote control modules are available and where are those modules located or designed for. With this info, the app may show a map to the user so that user knows where are the control modules of each type, where the user sits. The app may ask the user to choose user's seat location (at least driver or passenger for dismissal of the lock screen) for later use. The app may get this information from other means, for example scan a RFID tag, a NFC chip or a QrCode. The user can choose a RC module or modules for remote control. 

Reviewer: ```suggestion
- When a user launches a RC app from a phone, the app calls GetSystemCapabilites to get vehicle configurations, which includes how many rows, columns and levels are available, how many / what type / what id of remote control modules are available, and where those modules are located and/or designed for. With this info, the app may show a map to the user so that the user knows where the control modules of each type are, and where the user sits. The app may ask the user to choose their seat location (at least driver or passenger for dismissal of the lock screen) for later use. The app may get this information from other means, for example scan an RFID tag, an NFC chip, or a QrCode. The user can choose a RC module or modules for remote control. 
```

Author: updated

- `GetInteriorVehicleDataConsent` needs be extended to mobile api, so that an app can send a request to control/reserve a RC module after getting the SystemCapabilites and user's intent. 

- Once the SDL receives a `GetInteriorVehicleDataConsent`, `SetInteriorVehicleData` or `ButtonPress`, if the requested RC module is exclusive to a user (`share=false`), without asking the driver, SDL shall grant the access only if the user's grid equals to or is within module service area. (The reason behind it is that a user shall be able to control his own seat without asking driver's permission, it does not make sense if a pop up is shown on driver's screen). Otherwise, SDL rejects the request.

- If the requested RC module is shared among multiple users (`share=true`), all users located within the module's service area can potentially control the module without asking driver's permission.

>   - If the RC settings is auto deny, then only the first user can control the module. Whoever requests a free shared module can lock the module and use it until the user free the module. 

Reviewer: ```suggestion
  - If the RC setting is auto deny, then only the first user can control the module. Whoever requests a free shared module can lock the module and use it until the user frees the module. 
```

Author: updated

>   - If the RC settings is auto allow, then every one can control the module. 

Author: changed

>   - If the RC settings is ask driver, SDL shall grant or deny the request following the current resource allocation rules: if module is free, allow access; if module is used by another user or app, and current app is in HMI FULL, SDL shall trigger or forward `GetInteriorVehicleDataConsent` to HMI. HMI shows a popup so that the driver can give permission to the app for accessing the requested module. 

Author: changed

  - In order to reduce driver distractions, the default settings shall be auto allow.

- Granted permission is valid within ignition cycle. Reset after ignition, disable and re-enable remote control, app exit or app request to release.

> - A new RPC `ReleaseInteriorVehicleDataModule` is needed to indicate that the app is done with the RC module and no longer need access to it.

Reviewer: ```suggestion
- A new RPC `ReleaseInteriorVehicleDataModule` is needed to indicate that the app is done with the RC module and no longer needs access to it.
```

Author: changed

</function>

<function name="ReleaseInteriorVehicleDataModule" functionID="ReleaseInteriorVehicleDataModuleID" messagetype="response" since="5.x">
    <param name="resultCode" type="Result" platform="documentation" mandatory="true">
        <description>See Result</description>
        <element name="SUCCESS"/>
        <element name="INVALID_DATA"/>
        <element name="OUT_OF_MEMORY"/>
        <element name="TOO_MANY_PENDING_REQUESTS"/>
        <element name="APPLICATION_NOT_REGISTERED"/>
        <element name="GENERIC_ERROR"/>
        <element name="REJECTED"/>
        <element name="IGNORED"/>
        <element name="DISALLOWED"/>
        <element name="UNSUPPORTED_RESOURCE"/>
    </param>
    <param name="info" type="String" maxlength="1000" mandatory="false">
    </param>
    <param name="success" type="Boolean" platform="documentation" mandatory="true">
        <description> true if successful; false, if failed </description>
    </param>
</function>
```

## Potential downsides

Not aware from the author.

## Impacts on existing code

- New parameters will need to be updated within the RPC.

- SDL core and the mobile proxies will need updates to support the new parameters.

> - There is no policy change regarding module permission. Policy still control per module type.

Reviewer: ```suggestion
- There is no policy change regarding module permission. Policy still controls per module type.
```

Author: changed

## Alternatives considered

> There are some alternatives of how to define the new module `Id` parameter without introduce the grid system.  

Reviewer: ```suggestion
There are some alternatives of how to define the new module `Id` parameter without introducing the grid system.  
```

Author: changed

#### Option 1: Duplicate parameters within module control data

This option duplicates all the parameters within a remote control module type with a slightly different name. 

Take the `temperature` parameter in "single-zone" climate control as an example, we may have two parameters, a `driver side temperature` and a `passenger side temperature` for "dual-zone" climate control. Each parameter is for a climate zone.

In this option, there is no need to identify the module, as there is still only one module per module type. Instead of having one parameter per module and multiple modules, it has multiple parameters in one big module. Additional parameters are added for the new control items.

> This option may work for a few parameters. For example, temperature only. However, there may be separate controls for airflow mode and fan speed in each climate control zones. In seat control, each seat has its own complete set of parameters.

Author: changed

It is not scalable for many parameters.  

#### Option 2: OEM assigned ID

> In this option, instead of saying a module is located in a grid, an OEM assigns an ID to each module. The ID may be a UUID that does not have any meaning that can be understand by a user or an app. The ID may be a meaningful ID for the module, for example, a seat number assigned by the OEM or anything meaningful that can allow a normal user/driver/passenger associate the ID to the module with just common sense.

Author: updated

OEM module ID + module type together identify the module and potentially UUID alone identifies the module.

> Once implemented, no additional SDL work when new modules are added for existing module type.

Author: changed

Each OEM may have its own way to assign IDs. It may cause confusions to app developers. It is hard to develop an SDL enabled app across OEMs.

```xml
<struct name="ModuleId" since="5.x">
    <param name="uuid" type="String" minlength="1" maxlength="100" mandatory="true">
</struct>
```

#### Option 3: SDLC approved ID (commonly used name or location)

In this option, SDL defines a list of meaningful module IDs, which is a list of enumerations approved by SDLC. Since this list is unified across all SDLC members, there shall be the same understanding across the developers and OEMs.
Every time a new module is added, a new proposal needs to be approved, and there will be a version change in mobile API.

In this option, the module ID and the module type together identify a module.

There is a "default", “main" or "primary" module ID for each module type. This default module ID may be used if there is only one module per module type. This ID may be also used when there are multiple modules but one of them is the most commonly used.

> There is a "driver" module ID. It is potential for the modules that are designed or mostly used for a driver. For example, driver climate, driver seat, driver light, driver audio.

Reviewer: ```suggestion
There is a "driver" module ID. It is potentially for the modules that are designed or mostly used for a driver. For example, driver climate, driver seat, driver light, driver audio.
```

Author: updated

There is a "passenger" or "passengers" module ID. To tell which passenger from all possible passengers, it may have "front passenger" or "rear row passengers". To extend the idea, the name of a location or seat position may be introduced. For example, rear left (the 2nd row vehicle), rear right (2nd row vehicle), second row, third row, etc. It is a short version of "grid" system with only commonly used location defined.

Consider a normal passenger vehicle with 2~8 seats. There are one, two or three rows: driver/front row only; front row and rear row; or front-middle-back rows; or equivalently front row, second row, and third row. There are usually two or three columns in a row: driver-passenger sides, left-right sides or left-middle-right sides.

> For vehicles with more (9, 12, 15) seats, there may be four or five rows and four seats per row. It get complex to give a name to each seat location. The "grid" system will work better. However, it is not common to have a RC module specific to a seat in these vehicles.

Author: updated

> We listed the most commonly use name here. 

Author: updated

```xml
<enum name="ModuleId">
    <description>most commonly used names or locations for a RC module. </description>
    <element name="DEFAULT"/> <!--main, default, front middle, center -->
    <element name="DRIVER"/>
    <element name="FRONT_PASSENGER"/>
    <element name="SECOND_ROW"/> <!--second row middle or as a whole row -->
    <element name="SECOND_ROW_LEFT"/>
    <element name="SECOND_ROW_RIGHT"/>
    <element name="THIRD_ROW"/> <!-- third row middle or as a whole row -->
    <element name="THIRD_ROW_LEFT"/>
    <element name="THIRD_ROW_RIGHT"/>
</enum>
```

> Not all combinations are valid. OEM shall make sure they publish a valid combination of `ModuleId` for each `moduleType` in `XyzControlCapabilities`. In addition, mobile apps shall only use IDs that are published by the control capabilities.

Reviewer: ```suggestion
Not all combinations are valid. An OEM shall make sure they publish a valid combination of `ModuleId` for each `moduleType` in `XyzControlCapabilities`. In addition, mobile apps shall only use IDs that are published by the control capabilities.
```

Author: updated

#### Option 4: SDLC assigned "module" ID

> Similar to option 3, but module ID include module type info. It is possible to use module ID alone to identify a module. We listed the most commonly use name as below. 

Author: updated

```xml
<enum name="ModuleId">

>     <description>most commonly used RC modules</description>

Reviewer: I believe all of these descriptions should ready "...used when supporting only one..."

Author: updated

    <element name="DEFAULT_CLIMATE"/> <!--  used when support only one CLIMATE -->
    <element name="DEFAULT_RADIO"/>   <!--  used when support only one RADIO -->
    <element name="DEFAULT_AUDIO"/>   <!--  used when support only one AUDIO -->
    <element name="DEFAULT_HMI_SETTINGS"/> <!--  used when support only one HMI -->
    <element name="DEFAULT_LIGHT"/>   <!--  used when support only one LIGHT -->
    <element name="DEFAULT_SEAT"/>    <!--  used when support only one SEAT -->

>     <element name="DRIVER_CLIMATE"/>  <!--  same as DEFAULT_CLIMATE when support multiples -->

Reviewer: ```suggestion
    <element name="DRIVER_CLIMATE"/>  <!--  same as DEFAULT_CLIMATE when supporting multiple modules -->
```

Author: updated

    <element name="FRONT_PASSENGER_CLIMATE"/>

>     <element name="DRIVER_SEAT"/>     <!--  same as DEFAULT_SEAT when support multiples -->

Author: updated

- SDL shall enforce permission control. It may or may not use the local policy table depending on the implementation. The permissions are given by the user/driver via HMI, and information is passed to sdl core.

> - When a user launches a RC app from a phone, the app calls GetSystemCapabilites to get vehicle configurations, which includes how many rows, columns and levels are available, how many / what type / what id of remote control modules are available, and where  those modules are located and/or designed for. With this info, the app may show a map to the user so that user knows where the control modules of each type are, and where the user sits. The app may ask the user to choose user's seat location (at least driver or passenger for dismissal of the lock screen) for later use. The app may get this information from other means, for example scan a RFID tag, a NFC chip or a QrCode. The user can choose a RC module or modules for remote control. 

Reviewer: A few updates are still needed:

- Remove extra space in "...and where  those modules are located"
- Change "The app may ask the user to choose user's seat location" to "The app may ask the user to choose their seat location"

Author: updated
